File: android/app/src/main/java/com/vegettable/app/ui/home/HomeFragment.java
package com.vegettable.app.ui.home;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.vegettable.app.R;
import com.vegettable.app.model.ApiResponse;
import com.vegettable.app.model.ProductSummary;
import com.vegettable.app.network.ApiClient;
import com.vegettable.app.ui.adapter.ProductAdapter;
import com.vegettable.app.ui.detail.DetailActivity;
import com.vegettable.app.util.PrefsManager;

import com.vegettable.app.model.PaginatedResponse;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class HomeFragment extends Fragment implements ProductAdapter.OnItemClickListener {

    private RecyclerView rvProducts;
    private ProgressBar progressBar;
    private LinearLayout layoutError;
    private TextView tvError;
    private SwipeRefreshLayout swipeRefresh;
    private ChipGroup chipGroupCategory;

    private ProductAdapter adapter;
    private PrefsManager prefs;
    private String selectedCategory = null;

    // 分頁狀態
    private static final int PAGE_SIZE = 20;
    private int currentOffset = 0;
    private boolean hasMore = false;
    private boolean isLoadingMore = false;

    private static final String[][] CATEGORIES = {
            {"all", "全部"}, {"vegetable", "蔬菜"}, {"fruit", "水果"},
            {"fish", "漁產"}, {"poultry", "肉品"}, {"flower", "花卉"}, {"rice", "白米"}
    };

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_home, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        prefs = new PrefsManager(requireContext());

        rvProducts = view.findViewById(R.id.rv_products);
        progressBar = view.findViewById(R.id.progress_bar);
        layoutError = view.findViewById(R.id.layout_error);
        tvError = view.findViewById(R.id.tv_error);
        swipeRefresh = view.findViewById(R.id.swipe_refresh);
        chipGroupCategory = view.findViewById(R.id.chip_group_category);

        // 設定 RecyclerView
        LinearLayoutManager layoutManager = new LinearLayoutManager(requireContext());
        rvProducts.setLayoutManager(layoutManager);
        adapter = new ProductAdapter(this, prefs.getFavorites());
        adapter.setPriceUnit(prefs.getPriceUnit());
        adapter.setShowRetail(prefs.isShowRetailPrice());
        rvProducts.setAdapter(adapter);

        // Infinite scroll — 接近底部時自動載入下一頁
        rvProducts.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                if (dy <= 0) return; // 只在向下滾動時觸發
                int visibleCount = layoutManager.getChildCount();
                int totalCount = layoutManager.getItemCount();
                int firstVisible = layoutManager.findFirstVisibleItemPosition();

                if (hasMore && !isLoadingMore && (visibleCount + firstVisible + 5) >= totalCount) {
                    loadMoreProducts();
                }
            }
        });

        // 分類 Chips
        setupCategoryChips();

        // 下拉刷新
        swipeRefresh.setColorSchemeResources(R.color.primary);
        swipeRefresh.setOnRefreshListener(this::loadProducts);

        // 重試按鈕
        view.findViewById(R.id.btn_retry).setOnClickListener(v -> loadProducts());

        // 載入資料
        loadProducts();
    }

    private void setupCategoryChips() {
        for (String[] cat : CATEGORIES) {
            Chip chip = new Chip(requireContext());
            chip.setText(cat[1]);
            chip.setCheckable(true);
            chip.setTag(cat[0]);
            if ("all".equals(cat[0])) {
                chip.setChecked(true);
            }
            chip.setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (isChecked) {
                    String tag = (String) buttonView.getTag();
                    selectedCategory = "all".equals(tag) ? null : tag;
                    loadProducts();
                }
            });
            chipGroupCategory.addView(chip);
        }
    }

    private void loadProducts() {
        progressBar.setVisibility(View.VISIBLE);
        layoutError.setVisibility(View.GONE);
        currentOffset = 0;

        ApiClient.getInstance().getApi().getProductsPaginated(selectedCategory, 0, PAGE_SIZE)
                .enqueue(new Callback<ApiResponse<PaginatedResponse<ProductSummary>>>() {
                    @Override
                    public void onResponse(@NonNull Call<ApiResponse<PaginatedResponse<ProductSummary>>> call,
                                           @NonNull Response<ApiResponse<PaginatedResponse<ProductSummary>>> response) {
                        if (!isAdded()) return;
                        swipeRefresh.setRefreshing(false);
                        progressBar.setVisibility(View.GONE);

                        if (response.isSuccessful() && response.body() != null
                                && response.body().isSuccess() && response.body().getData() != null) {
                            PaginatedResponse<ProductSummary> page = response.body().getData();
                            List<ProductSummary> products = page.getItems();
                            adapter.setItems(products);
                            hasMore = page.isHasMore();
                            currentOffset = products.size();
                            rvProducts.setVisibility(View.VISIBLE);

                            // 快取
                            prefs.cacheProducts(new Gson().toJson(products));
                        } else {
                            showError("無法取得資料");
                            loadCachedProducts();
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<ApiResponse<PaginatedResponse<ProductSummary>>> call,
                                          @NonNull Throwable t) {
                        if (!isAdded()) return;
                        swipeRefresh.setRefreshing(false);
                        progressBar.setVisibility(View.GONE);
                        showError("網路錯誤: " + t.getMessage());
                        loadCachedProducts();
                    }
                });
    }

    private void loadMoreProducts() {
        if (!hasMore || isLoadingMore) return;
        isLoadingMore = true;

        ApiClient.getInstance().getApi().getProductsPaginated(selectedCategory, currentOffset, PAGE_SIZE)
                .enqueue(new Callback<ApiResponse<PaginatedResponse<ProductSummary>>>() {
                    @Override
                    public void onResponse(@NonNull Call<ApiResponse<PaginatedResponse<ProductSummary>>> call,
                                           @NonNull Response<ApiResponse<PaginatedResponse<ProductSummary>>> response) {
                        isLoadingMore = false;
                        if (!isAdded()) return;

                        if (response.isSuccessful() && response.body() != null
                                && response.body().isSuccess() && response.body().getData() != null) {
                            PaginatedResponse<ProductSummary> page = response.body().getData();
                            List<ProductSummary> newItems = page.getItems();
                            adapter.addItems(newItems);
                            hasMore = page.isHasMore();
                            currentOffset += newItems.size();
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<ApiResponse<PaginatedResponse<ProductSummary>>> call,
                                          @NonNull Throwable t) {
                        isLoadingMore = false;
                    }
                });
    }

    private void loadCachedProducts() {
        String cached = prefs.getCachedProducts();
        if (cached != null) {
            try {
                List<ProductSummary> products = new Gson().fromJson(cached,
                        new TypeToken<List<ProductSummary>>(){}.getType());
                if (products != null && !products.isEmpty()) {
                    adapter.setItems(products);
                    rvProducts.setVisibility(View.VISIBLE);
                    layoutError.setVisibility(View.GONE);
                }
            } catch (Exception ignored) {}
        }
    }

    private void showError(String message) {
        tvError.setText(message);
        layoutError.setVisibility(View.VISIBLE);
    }

    @Override
    public void onItemClick(ProductSummary product) {
        Intent intent = new Intent(requireContext(), DetailActivity.class);
        intent.putExtra("cropName", product.getCropName());
        intent.putExtra("cropCode", product.getCropCode());
        startActivity(intent);
    }

    @Override
    public void onFavoriteClick(ProductSummary product) {
        prefs.toggleFavorite(product.getCropCode());
        // 只更新被點擊的項目，而非整個列表
        adapter.notifyItemRangeChanged(0, adapter.getItemCount());
    }

    @Override
    public void onResume() {
        super.onResume();
        // 更新收藏狀態和設定
        if (adapter != null) {
            adapter.setPriceUnit(prefs.getPriceUnit());
            adapter.setShowRetail(prefs.isShowRetailPrice());
        }
    }
}

File: android/app/src/main/java/com/vegettable/app/ui/home/HomeFragment.kt
package com.vegettable.app.ui.home

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.google.android.material.chip.Chip
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.vegettable.app.R
import com.vegettable.app.databinding.FragmentHomeBinding
import com.vegettable.app.model.ApiResponse
import com.vegettable.app.model.ProductSummary
import com.vegettable.app.network.ApiClient.Companion.instance
import com.vegettable.app.ui.adapter.ProductAdapter
import com.vegettable.app.ui.adapter.SkeletonAdapter
import com.vegettable.app.ui.detail.DetailActivity
import com.vegettable.app.util.PrefsManager
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class HomeFragment : Fragment(), ProductAdapter.OnItemClickListener {
    private var binding: FragmentHomeBinding? = null
    private var adapter: ProductAdapter? = null
    private var prefs: PrefsManager? = null
    private var selectedCategory: String? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentHomeBinding.inflate(inflater, container, false)
        return binding!!.getRoot()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        prefs = PrefsManager(requireContext())

        setupRecyclerView()
        setupCategoryChips()
        setupListeners()

        loadProducts()
    }

    private fun setupRecyclerView() {
        binding!!.rvProducts.setLayoutManager(LinearLayoutManager(requireContext()))
        adapter = ProductAdapter(this, prefs!!.favorites)
        adapter!!.setPriceUnit(prefs!!.priceUnit)
        adapter!!.setShowRetail(prefs!!.isShowRetailPrice)
        binding!!.rvProducts.setAdapter(adapter)

        // 骨架屏
        binding!!.rvSkeleton.setLayoutManager(LinearLayoutManager(requireContext()))
        binding!!.rvSkeleton.setAdapter(SkeletonAdapter(8))
    }

    private fun setupListeners() {
        // 下拉刷新配色與事件
        binding!!.swipeRefresh.setColorSchemeResources(R.color.primary)
        binding!!.swipeRefresh.setOnRefreshListener(OnRefreshListener { this.loadProducts() })

        // 重試按鈕
        binding!!.btnRetry.setOnClickListener(View.OnClickListener { v: View? -> loadProducts() })
    }

    private fun setupCategoryChips() {
        binding!!.chipGroupCategory.removeAllViews()
        for (cat in CATEGORIES) {
            // 使用 ContextThemeWrapper 或指定 Style
            val chip = Chip(requireContext(), null, com.google.android.material.R.attr.chipStyle)
            chip.setText(cat!![1])
            chip.setCheckable(true)
            chip.setTag(cat[0])

            // 套用你的 Liquid Glass 膠囊樣式
            // 注意：如果 XML 中定義了 CategoryChip style，建議在布局中動態引用
            if ("all" == cat[0]) {
                chip.setChecked(true)
            }

            chip.setOnCheckedChangeListener(CompoundButton.OnCheckedChangeListener { buttonView: CompoundButton?, isChecked: Boolean ->
                if (isChecked) {
                    val tag = buttonView!!.getTag() as String?
                    selectedCategory = if ("all" == tag) null else tag
                    loadProducts()
                }
            })
            binding!!.chipGroupCategory.addView(chip)
        }
    }

    private fun loadProducts() {
        // 如果不是下拉刷新觸發的，顯示骨架屏
        if (!binding!!.swipeRefresh.isRefreshing()) {
            binding!!.rvSkeleton.setVisibility(View.VISIBLE)
            binding!!.rvProducts.setVisibility(View.GONE)
        }
        binding!!.layoutError.setVisibility(View.GONE)

        instance!!.api.getProducts(selectedCategory)!!
            .enqueue(object : Callback<ApiResponse<MutableList<ProductSummary?>?>?> {
                override fun onResponse(
                    call: Call<ApiResponse<MutableList<ProductSummary?>?>?>,
                    response: Response<ApiResponse<MutableList<ProductSummary?>?>?>
                ) {
                    if (!isAdded()) return

                    binding!!.swipeRefresh.setRefreshing(false)
                    binding!!.rvSkeleton.setVisibility(View.GONE)

                    if (response.isSuccessful() && response.body() != null && response.body()!!.isSuccess && response.body()!!.data != null) {
                        val products: MutableList<ProductSummary> = response.body()!!.data!!.filterNotNull().toMutableList()
                        adapter!!.setItems(products)
                        binding!!.rvProducts.setVisibility(View.VISIBLE)
                        binding!!.tvOfflineBanner.visibility = View.GONE

                        // 快取最新的產品列表
                        prefs!!.cacheProducts(Gson().toJson(products))
                    } else {
                        val errorMsg = when (response.code()) {
                            400 -> "請求格式錯誤"
                            404 -> "查無資料"
                            429 -> "請求過於頻繁，請稍後再試"
                            in 500..599 -> "伺服器發生錯誤，請稍後再試"
                            else -> "無法取得最新資料"
                        }
                        handleLoadError(errorMsg)
                    }
                }

                override fun onFailure(
                    call: Call<ApiResponse<MutableList<ProductSummary?>?>?>,
                    t: Throwable
                ) {
                    if (!isAdded()) return

                    binding!!.swipeRefresh.setRefreshing(false)
                    binding!!.rvSkeleton.setVisibility(View.GONE)
                    handleLoadError("網路連線不穩定")
                }
            })
    }

    private fun handleLoadError(message: String?) {
        // 嘗試從快取讀取
        val cached = prefs!!.cachedProducts
        if (cached != null) {
            try {
                val products = Gson().fromJson<MutableList<ProductSummary>>(
                    cached,
                    object : TypeToken<MutableList<ProductSummary>>() {}.getType()
                )
                if (products != null && !products.isEmpty()) {
                    adapter!!.setItems(products)
                    binding!!.rvProducts.setVisibility(View.VISIBLE)
                    val ageText = prefs!!.cacheAgeText
                    binding!!.tvOfflineBanner.text = "⚠️ 離線模式｜顯示 $ageText 的快取資料"
                    binding!!.tvOfflineBanner.visibility = View.VISIBLE
                    return
                }
            } catch (ignored: Exception) {
            }
        }

        // 若連快取都沒有，則顯示錯誤畫面
        binding!!.tvError.setText(message)
        binding!!.layoutError.setVisibility(View.VISIBLE)
        binding!!.rvProducts.setVisibility(View.GONE)
    }

    override fun onItemClick(product: ProductSummary?) {
        product?.let {
            val intent = Intent(requireContext(), DetailActivity::class.java)
            intent.putExtra("cropName", it.cropName)
            intent.putExtra("cropCode", it.cropCode)
            startActivity(intent)
        }
    }

    override fun onFavoriteClick(product: ProductSummary?) {
        product?.let {
            prefs!!.toggleFavorite(it.cropCode)
            // 更新 Adapter 中的收藏狀態
            adapter!!.setFavorites(prefs!!.favorites)
            adapter!!.notifyDataSetChanged()
        }
    }

    override fun onResume() {
        super.onResume()
        // 當從 DetailActivity 返回時，確保收藏狀態與單位設定同步更新
        if (adapter != null) {
            adapter!!.setPriceUnit(prefs!!.priceUnit)
            adapter!!.setShowRetail(prefs!!.isShowRetailPrice)
            adapter!!.setFavorites(prefs!!.favorites) // 關鍵：同步收藏狀態
            adapter!!.notifyDataSetChanged()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null // 避免 Memory Leak
    }

    companion object {
        private val CATEGORIES = arrayOf<Array<String?>?>(
            arrayOf<String?>("all", "全部"),
            arrayOf<String?>("vegetable", "蔬菜"),
            arrayOf<String?>("fruit", "水果"),
            arrayOf<String?>("fish", "漁產"),
            arrayOf<String?>("poultry", "肉品"),
            arrayOf<String?>("flower", "花卉"),
            arrayOf<String?>("rice", "白米")
        )
    }
}
